gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/reactive/handlers/api/security/plan/AbstractSecurityPlan.java

```diff
@@ -0,0 +1,161 @@
+ * Security plan is responsible to
+ * <ul>
+ *     <li>Check if a policy can handle the security or not</li>
+ *     <li>Check the eventual selection rule matches (useful when dealing with multiple plans relying on the same security scheme such as <code>Authorization: bearer xxx</code> for JWT and OAuth)</li>
+ *     <li>Check if the policy requires to validate there is an associated subscription or not and validate the subscription accordingly</li>
+ *     <li>Invoke the <code>onSubscriptionInvalid</code> method when necessary</li>
+ * </ul>
+ *
+ * @author Jeoffrey HAEYAERT (jeoffrey.haeyaert at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+@Slf4j
+public abstract class AbstractSecurityPlan<T extends BaseSecurityPolicy, C extends BaseExecutionContext> {
+
+    protected static final Maybe<Boolean> TRUE = Maybe.just(true);
+    protected static final Maybe<Boolean> FALSE = Maybe.just(false);
+    protected final T policy;
+    private final String planId;
+    private final String selectionRule;
+
+    public AbstractSecurityPlan(@Nonnull final String planId, @Nonnull final T policy, final String selectionRule) {
+        this.planId = planId;
+        this.policy = policy;
+        this.selectionRule = getSelectionRule(selectionRule);
+    }
+
+    public String id() {
+        return policy.id();
+    }
+
+    /**
+     * Make sure the security plan can be executed for the current request.
+     *
+     * @param ctx the current execution context.
+     * @return <code>true</code> if this security plan can be executed for the request, <code>false</code> otherwise.
+     */
+    public Single<Boolean> canExecute(C ctx) {
+        return extractSecurityToken(ctx)
+            .flatMap(securityToken -> {
+                ctx.setInternalAttribute(ATTR_INTERNAL_SECURITY_TOKEN, securityToken);
+                if (!securityToken.isInvalid()) {
+                    return isApplicableWithValidSubscription(ctx, securityToken);
+                }
+                return Maybe.empty();
+            })
+            .defaultIfEmpty(false);
+    }
+
+    /**
+     * Invokes the policy's method for security.
+     *
+     * @param ctx the current execution context.
+     * @return a {@link Completable} that completes when the security policy has been successfully executed or returns an error otherwise.
+     */
+    public Completable execute(final C ctx, final ExecutionPhase executionPhase) {
+        return executeSecurityPolicy(ctx, executionPhase).doOnSubscribe(disposable -> ctx.setAttribute(ATTR_PLAN, planId));
+    }
+
+    public int order() {
+        return policy.order();
+    }
+
+    protected abstract Maybe<SecurityToken> extractSecurityToken(final C executionContext);
+
+    protected abstract Completable executeSecurityPolicy(final C ctx, final ExecutionPhase executionPhase);
+
+    protected abstract String getSelectionRule(String selectionRule);
```

Comment:
Just to given another point of view. Another way to do it can be to create `SecurityPlan<T extends BaseSecurityPolicy, C extends BaseExecutionContext>` that take a parameter an object that respect the interface with this 3 methods.
This another way is a composition and is generally present better than inheritence.

But I don't know all history and context of this piece of code